### ISSUE TITLE
label:question Notebook Chapter 4

### ISSUE THREAD

**Issue by anu43** [ageron/handson-ml2#211]
In Chapter 4 notebook block 37, I wonder which method between fit and fit_transform is used by the pipeline for PolynomialFeatures and how? I mean according to the previous blocks it is clear that we need to use fit_transform but I could not see a specific indication of it within the code.

```python
polynomial_regression = Pipeline([
    ('poly_features', PolynomialFeatures(degree=10, include_bias=False)),
    ('lin_reg', LinearRegression()),
])
```

**Comment by Dseal95** [comment]
Hi Anu, 

If you look at the pipeline documentation on **scikit-learn, https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html** you can see that a pipeline is simply a list of transforms and a final estimator. This means that the final step in the pipeline (linear regression() in this instance) is the estimator and then the preceding steps (polynomial features() in this instance) are all transforms to some input data. T**he pipeline thus uses fit_transform for the transformers preceding the model object (estimator)**

For the example above, if you added the following piece of code:

poly = polynomial_regression.fit(x_train, y_train)

The sequential steps that the code would execute are:

1. The pipeline will take some data, X containing n features and then transform the data (using fit_transform) using polynomial features() object. This would transform the n features in X to (n+d)!/n!d! number of features where d = the degree of freedom (d=10 here). 
2. the transformed data would then be fitted to the Lin_reg step of the pipeline. You can then use the poly (taking the code above) to predict scores for any new data (validation or test) 

The reason for this is the Fit() method will fit all the transforms one after the other and transform the data, then fit the transformed data using the final estimator.

You can go one step further and add another transformer, say a StandardScalar() object after the polynomial transform and this would just add a transformation step. The model will then not only be fitted to poly transformed data but scaled, polynomial transformed data. 

If you want to test the way the pipeline works, you can break up the code and test the two outputs on the same validation data. Considering, input training data X_train and y_train:

**1. Method1:**
polynomial_regression = Pipeline([
    ('poly_features', PolynomialFeatures(degree=10, include_bias=False)),
    ('lin_reg', LinearRegression()),
])

Lin_reg = polynomial_regression.fit(X_train, y_train)

**2. Method2:**
poly= PolynomialFeatures(degree=10, include_bias=False)
poly_features = poly.fit_transform(X_train) 

lin_reg = LinearRegression()
lin_reg = lin_reg.fit(poly_features, y_train)

you can then compare the 
lin_reg.intercept_
lin_reg.coef_
Or its predictions on some new data. 

Note: you will need to stay consistent with a random_state hyperparam for results to match. 

I hope that helps.

Regards, 
Dan

**Comment by JosephChan101** [comment]
The pipeline works does the following:

Step 1: poly_features = PolynomialFeatures(degree=10, include_bias=False)),
Step 2: X_transform = poly_features.fit_transform(X_train)
Step 3: y_transform = poly_features.fit_transform(y_train)
Step 4: lin_reg = LinearRegression()
Step 5: lin_reg.fit(X_transform, y_transform)

When you call polynomial_regression.fit(X_train, y_train), the pipeline will go through those steps sequentially. So to answer your question, the pipeline use fit_transform() instead of fit()

**Comment by anu43** [comment]
Thanks, @Dseal95 @JosephChan101 for sparing some time for answering!
